### PR TITLE
[tests] Add integration tests for array data type

### LINKED ISSUE #441: Add integration tests for array support in Fluss
### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss-rust/issues) and found nothing similar.

### Description

Add integration tests for array support in Fluss

### Willingness to contribute

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by charlesdong1991**
### Purpose

Add integration tests for array data type support in Rust core and Python bindings. 

c++ has integration tests covered already, so only focus on rust and python

Linked issue: close #441 

### Tests

All tests have passed locally.

### CODE REVIEW

File: bindings/python/test/test_kv_table.py [hunk 1]
```diff
@@ -335,6 +346,228 @@ async def test_partitioned_table_upsert_and_lookup(connection, admin):
...
+    schema = fluss.Schema(
+        pa.schema(
+            [
+                pa.field("id", pa.int32()),
+                pa.field("arr_long_str", pa.list_(pa.string())),
+                pa.field("arr_big_decimal", pa.list_(pa.decimal128(22, 5))),
+                pa.field("arr_ts_nano", pa.list_(pa.timestamp("ns"))),
+                pa.field("arr_float", pa.list_(pa.float32())),
+                pa.field("arr_double", pa.list_(pa.float64())),
+                # TODO(fluss-python): support PyArrow FixedSizeBinary in schema conversion.
```
Comment (charlesdong1991) [inline bindings/python/test/test_kv_table.py]: i seems fixed size binary will fail now, and should be supported.

it is nicer to address in a separate PR IMHO, so added a new issue: https://github.com/apache/fluss-rust/issues/524

**Comment by charlesdong1991** [comment]
@leekeiabstraction @fresh-borzoni it will be great if you can take a look! 🙏

File: crates/fluss/tests/integration/kv_table.rs [hunk 2]
```diff
@@ -20,14 +20,41 @@
...
-    use fluss::row::{GenericRow, InternalRow};
+    use fluss::row::binary_array::FlussArrayWriter;
+    use fluss::row::{FlussArray, GenericRow, InternalRow};
 
     fn make_key(id: i32) -> GenericRow<'static> {
-        let mut row = GenericRow::new(3);
+        make_key_with_len(id, 3)
+    }
+
+    fn make_key_with_len(id: i32, field_count: usize) -> GenericRow<'static> {
```
Comment (fresh-borzoni) [inline crates/fluss/tests/integration/kv_table.rs]: nit: `make_key_with_field_count` feels like a better name

File: crates/fluss/tests/integration/log_table.rs [hunk 3]
```diff
@@ -1373,4 +1388,546 @@ mod table_test {
...
+    #[tokio::test]
+    async fn append_and_scan_with_array_encoding_edge_cases() {
+        use fluss::row::binary_array::FlussArrayWriter;
+        use fluss::row::{Decimal, FlussArray, GenericRow, TimestampNtz};
+
+        fn assert_f32_special(actual: f32, expected: f32) {
+            if expected.is_nan() {
+                assert!(actual.is_nan(), "expected NaN");
+            } else if expected.is_infinite() {
+                assert_eq!(actual.is_infinite(), true);
```
Comment (fresh-borzoni) [inline crates/fluss/tests/integration/log_table.rs]: nit: assert!(actual.is_infinite()) ?

File: crates/fluss/tests/integration/log_table.rs [hunk 4]
```diff
@@ -1373,4 +1388,546 @@ mod table_test {
...
+            } else {
+                assert!((actual - expected).abs() < f32::EPSILON);
+            }
+        }
+
+        fn assert_f64_special(actual: f64, expected: f64) {
+            if expected.is_nan() {
+                assert!(actual.is_nan(), "expected NaN");
+            } else if expected.is_infinite() {
+                assert_eq!(actual.is_infinite(), true);
```
Comment (fresh-borzoni) [inline crates/fluss/tests/integration/log_table.rs]: ditto

File: crates/fluss/tests/integration/kv_table.rs [hunk 5]
```diff
@@ -20,14 +20,41 @@
...
+        let mut row = GenericRow::new(field_count);
         row.set_field(0, id);
         row
     }
 
+    fn make_string_array(values: &[Option<&str>]) -> FlussArray {
```
Comment (fresh-borzoni) [inline crates/fluss/tests/integration/kv_table.rs]: we have a lot of similar helpers shared between log_table.rs and kv_table.rs
Let's factor them out to utils.rs, wdyt?

File: Cargo.lock [hunk 6]
```diff
@@ -746,6 +746,15 @@ dependencies = [
  "rustc_version",
 ]
 
+[[package]]
```
Comment (fresh-borzoni) [inline Cargo.lock:749]: we can rebase now with Cargo.lock synced

**Review by fresh-borzoni — APPROVED** [review]
@charlesdong1991 Ty for the PR, LGTM overall
Left a couple of nits and cleanups, non-blocking though

**Comment by charlesdong1991** [comment]
Thanks for your comments, addressed all, and rebased! @fresh-borzoni PTAL 🙏

**Comment by fresh-borzoni** [comment]
> Thanks for your comments, addressed all, and rebased! @fresh-borzoni PTAL 🙏

All good, thank you for addressing 👍

File: crates/fluss/tests/integration/kv_table.rs [hunk 7]
```diff
@@ -880,6 +896,226 @@ mod kv_table_test {
...
+        );
+
+        admin
+            .drop_table(&table_path, false)
+            .await
+            .expect("Failed to drop table");
+    }
+
+    #[tokio::test]
+    async fn upsert_and_lookup_with_array() {
```
Comment (leekeiabstraction) [inline crates/fluss/tests/integration/kv_table.rs:911]: Hmm, this approach increases the number of test cases (eventually when we have map types support as well). Can you try to incorporate these into `all_supported_datatypes`? You can have additional columns there e.g. array of arrays.

File: crates/fluss/tests/integration/log_table.rs [hunk 8]
```diff
@@ -1373,4 +1384,523 @@ mod table_test {
             .await
             .expect("Failed to drop table");
     }
+
+    #[tokio::test]
+    async fn append_and_scan_with_array() {
```
Comment (leekeiabstraction) [inline crates/fluss/tests/integration/log_table.rs:1389]: Similarly here on incorporating into `all_supported_datatypes` test case.

**Review by leekeiabstraction — COMMENTED** [review]
TY for the PR. left a couple of comments. PTAL!

File: crates/fluss/tests/integration/kv_table.rs (hunk 7 above)
Comment (charlesdong1991) [reply crates/fluss/tests/integration/kv_table.rs:911]: hmm, array test coverage is also added to `all_supported_datatypes` for both logs and kv tests. This test cover the patterns that don't fit `all_supported_datetypes` structure that well IMHO, since these require multi-row setups and different null patterns per row. 

And if i understand correctly, `all_supported_datetypes` tests one row with values and one row with all nulls.

If we add Map, i think it's best to follow the same approach as here for Array, so basic column incorporated in `all_supported_datatypes` while specific tests for map-specific edge cases.

But i don't have very strong opinion on this, happy to adjust if you prefer the consolidation! @leekeiabstraction 🙏

**Comment by charlesdong1991** [comment]
i think https://github.com/apache/fluss-rust/actions/runs/25275194644/job/74108466204 is a test flake, and shouldn't be caused by changes in this pr @leekeiabstraction @fresh-borzoni hope rerun will fix.

in the meantime, i can take a quick look to understand why

File: crates/fluss/tests/integration/log_table.rs (hunk 8 above)
Comment (charlesdong1991) [reply crates/fluss/tests/integration/log_table.rs:1389]: similar reply above 🙏

**Comment by fresh-borzoni** [comment]
> i think https://github.com/apache/fluss-rust/actions/runs/25275194644/job/74108466204 is a test flake, and shouldn't be caused by changes in this pr @leekeiabstraction @fresh-borzoni hope rerun will fix.
> 
> in the meantime, i can take a quick look to understand why

@charlesdong1991 Yes, it's just flake with python tests, we use single test-cluster across 4 different python version x 4 workers, so under contention this can happen. I need to take a look at this, because it's getting annoying sometimes :)

File: crates/fluss/tests/integration/kv_table.rs (hunk 7 above)
Comment (leekeiabstraction) [reply crates/fluss/tests/integration/kv_table.rs:911]: Thanks Charles for the response. Let's do this, leave `all_supported_datatypes` as you have already updated and then consolidate the array / nested array and potentially the edge case test cases into a single `upsert_and_lookup_with_array ` test case. That way we can leave `all_supported_datatypes` simple and test cases organised in a way that is mostly focused on features of primary key table.

File: crates/fluss/tests/integration/log_table.rs (hunk 8 above)
Comment (leekeiabstraction) [reply crates/fluss/tests/integration/log_table.rs:1389]: Similar reply to the above.

**Comment by charlesdong1991** [comment]
Consolidated the tests, PTAL! @leekeiabstraction thanks!

File: bindings/python/test/test_kv_table.py [hunk 9]
```diff
@@ -335,6 +346,228 @@ async def test_partitioned_table_upsert_and_lookup(connection, admin):
...
+
+    result3 = await lookuper.lookup({"id": 3})
+    assert result3 is not None
+    assert result3["tags"] is None
+    assert result3["scores"] == [42]
+
+    await admin.drop_table(table_path, ignore_if_not_exists=False)
+
+
+async def test_upsert_and_lookup_with_nested_array(connection, admin):
```
Comment (leekeiabstraction) [inline bindings/python/test/test_kv_table.py]: Can we merge this test case with above ask well?

File: bindings/python/test/test_kv_table.py [hunk 10]
```diff
@@ -335,6 +346,228 @@ async def test_partitioned_table_upsert_and_lookup(connection, admin):
...
+    assert result3["matrix"] is None
+
+    result4 = await lookuper.lookup({"id": 4})
+    assert result4 is not None
+    assert result4["matrix"] == [[1, None], None, []]
+
+    await admin.drop_table(table_path, ignore_if_not_exists=False)
+
+
+async def test_upsert_and_lookup_with_array_rich_types(connection, admin):
```
Comment (leekeiabstraction) [inline bindings/python/test/test_kv_table.py:422]: Same here.

File: bindings/python/test/test_kv_table.py [hunk 11]
```diff
@@ -335,6 +346,228 @@ async def test_partitioned_table_upsert_and_lookup(connection, admin):
...
+    assert result["arr_ts_ntz"] == [datetime(2026, 1, 23, 10, 13, 47, 123000)]
+    assert result["arr_ts_ltz"] == [
+        datetime(2026, 1, 23, 10, 13, 47, 123000, tzinfo=timezone.utc)
+    ]
+    assert result["arr_decimal"] == [Decimal("123.45"), None]
+
+    await admin.drop_table(table_path, ignore_if_not_exists=False)
+
+
+async def test_upsert_and_lookup_with_array_encoding_edge_cases(connection, admin):
```
Comment (leekeiabstraction) [inline bindings/python/test/test_kv_table.py]: Same here. IMO we can reduce / simplify some of these by merging the array tests into a single test case.

File: crates/fluss/tests/integration/log_table.rs [hunk 12]
```diff
@@ -1373,4 +1384,430 @@ mod table_test {
...
+        assert!(r2.is_null_at(3).unwrap());
+
+        admin
+            .drop_table(&table_path, false)
+            .await
+            .expect("Failed to drop table");
+    }
+
+    #[tokio::test]
+    async fn append_and_scan_with_array_rich_types() {
```
Comment (leekeiabstraction) [inline crates/fluss/tests/integration/log_table.rs:1524]: Same here.

**Review by leekeiabstraction — COMMENTED** [review]
Left further comments, PTAL @charlesdong1991 🙏

**Comment by charlesdong1991** [comment]
Consolidated @leekeiabstraction PTAL 🙏

**Comment by fresh-borzoni** [comment]
@charlesdong1991 I think it's good enough for now, merging 👍 

We have an inconsistency between bindings' tests though, as we have a split: structural tests and rich types test in essence, but C++ and Rust don't have this for KV tables, while Python has this for both KV and log. 

@leekeiabstraction Your point on test sprawl is fair - Would you mind to file a couple of 'first issues' to consolidate tests and make them more consistent and manageable? Happy to take them myself if you'd rather not.
